design-proposals/klip-35-dynamic-processing-log-levels.md

```diff
@@ -0,0 +1,99 @@
+
+There have also been incidents where even the current error-only level of logging is too verbose. In such cases, it should be possible to turn the processing log off at runtime.
+
+In this KLIP, I propose a mechanism to dynamically control the log level of the record-processing log. Furthermore, the proposed mechanism makes it possible to change the level with a fine granularity - so you can change the level at runtime for individual queries, or even individual processing nodes of a query.
+
+For important background on the processing log, please read https://docs.ksqldb.io/en/latest/developer-guide/test-and-debug/processing-log/
+
+## In-Scope
+
+- API for getting/setting the global log level
+- API for getting/setting query/node-level log levels
+- Kafka-based stateful implementation that distributes level settings to the cluster and stores them so they are recovered on a restart.
+
+## Out-Of-Scope
+
+- Thoroughly populating the query engine with log messages to address the problems we’ve already seen. This is super important, but out of scope for this particular KLIP.
+
+## Value/Return
+
+A step toward improving UX when users observe unexpected behavior from running queries.
+
+## Public APIs
+
+### System Table for Getting/Setting Log Levels
+
+The table will have the following schema:
+
+```
+LEVEL STRING (one of <OFF|FATAL|ERROR|WARN|INFO|DEBUG>)
+```
+
+## Design
+
+The system table's implementation will work as follows:
+
+### Storing Log Levels
+
+The log levels will be stored on a Kafka topic called the processing log config topic (PLCT). The key of the topic’s records will be the processing logger name. We will use the name “root” for the root logger. The value of the topic will be a JSON object representing the logger’s config. For now, it will contain a single field called “level”:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+```
+
+The Kafka topic will be compacted.
+
+The name of the topic will be controlled by a configuration named “ksql.logging.processing.config.topic.name”.
+
+### Setting a Log Level
+
+To set the log level, the ksql node receiving the INSERT request to set the level will produce a record with the logger’s key and value containing the specified level to the PLCT. Each ksql node will run a thread that consumes the PLCT. For each message consumed, the thread will use the log4j API to set the log level. To set the root level, the thread will call:
+
+```
+LogManager.getRootLogger().setLevel(<level>);
+```
+
+To set a child logger level, the thread will call:
+
+```
+LogManager.getLogger().setLevel(<level>);
+```
+
+Unsetting a Log Level
+
+To unset the log level, the thread consuming the PLCT will set the log level for the specified logger to `null`.
```

Comment:
since we don't support `DELETE FROM` (yet) how will the end user issue something to reset the log level? `INSERT INTO plct (logger, level) VALUES ('id', null);`?